src/common/mempool.cc

```diff
@@ -12,21 +12,73 @@
  *
  */
 
+#include <thread>
 #include "include/mempool.h"
 #include "include/demangle.h"
 
-#if defined(_GNU_SOURCE) && defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
-#else
-// Thread local variables should save index, not &shard[index],
-// because shard[] is defined in the class
-static thread_local size_t thread_shard_index = mempool::num_shards;
-#endif
-
 // default to debug_mode off
 bool mempool::debug_mode = false;
 
 // --------------------------------------------------------------
 
+namespace mempool {
+
+static size_t num_shard_bits;
+static size_t num_shards;
+
+static thread_local size_t thread_shard_index = max_shards;
+
+#if !defined(MEMPOOL_SCHED_GETCPU)
+// There is no cheap way of determining which core a thread is assigned to.
+// Statically assign threads to shards to minimize how many threads use the
+// same shard
+static int thread_shard_next;
+static std::mutex thread_shard_mtx;
+
+#else
+// Use sched_getcpu() to determine which core a thread is running on
+#endif
+
+}
+
+size_t mempool::get_num_shards(void) {
+  if (num_shards==0) {
+    size_t threads = std::thread::hardware_concurrency();
+    if (threads == 0) {
+      threads = default_shards;
+    }
+    if (threads < min_shards) {
+      threads = min_shards;
+    }
+    if (threads > max_shards) {
+      threads = max_shards;
+    }
+    threads--;
+    while (threads!=0) {
+      num_shard_bits++;
+      threads>>=1;
+    }
+    num_shards = 1 << num_shard_bits;
+  }
+  return num_shards;
+}
+
+int mempool::pick_a_shard_int(void) {
+#if !defined(MEMPOOL_SCHED_GETCPU)
+  if (thread_shard_index == max_shards) {
+    // Thread has not been assigned to a shard yet
+    std::lock_guard<std::mutex> lck (thread_shard_mtx);
+    thread_shard_index = thread_shard_next++ & ((1 << num_shard_bits) - 1);
+  }
+#else
+  if (thread_shard_index == max_shards) {
+    // Thread has not been assigned to a shard yet
+    thread_shard_index = sched_getcpu() & ((1 << num_shard_bits) - 1);
+  }
+#endif
+  return thread_shard_index;
```

Comment:
I've updated the benchmarks in the PR comments. A local-exec TLS lookup is the cheapest at ~ 5 cycles, a global-dynamic TLS lookup (which will happen for code in a dynamic library so is more representative) is ~15 cycles, calling sched_getcpu() is ~32 cycles.

The call to pthread_self() and hash is ~8 cycles, this is effectively a local-exec TLS lookup and a bit of overhead for a function call and a logical and.

Inspecting the assembly code for the sched_getcpu() code path is over twice as long as a global-dynamic TLS lookup involving a call to __tls_get_addr() which confirms what the benchmarks are showing. Like you, I'm surprised sched_getcpu() isn't more efficient - but there's quite a bit of complexity working out whether it can use the VDSO variable or has to resort to making a syscall and this happens on every call.

I think we need to look at the mempool statistics when running on a system with a heavy I/O workload. The stats I've managed to obtain from running on a fairly pathetic setup show a 3x improvement for the adaptive caching versus calling sched_getcpu(), which might reduce to 2x if we made a slick inline version of pick_a_shard_int that just called sched_getcpu(). I suspect the benefits will be much greater on a system running heavy I/O workload, if they are not then maybe the simplicity of always calling sched_getcpu() will out weight a marginal performance improvement of trying to cache the result.

Reply:
Latest commit https://github.com/ceph/ceph/pull/55696/commits/5f721120ab8433689b6c1601077c17f3755b471c implements the following algorithms:

THREAD_OWNER - this has one TLS access per mempool stat update, only calls sched_getcpu() when a thread has context switched to another thread, and almost eliminates cache lines storing the stats from being accessed by different CPU cores (there is one cache line ping pong each time a thread switches cores).

SCHED_GETCPU - this calls sched_getcpu() every mempool stat update, it has no TLS access. Even with the optimized version of sched_getcpu() reading a variable from vDSO this slower than the THREAD_OWNER implementation.

SCHED_GETCPU USING RDPID - for systems that support rdpid, use this instead of sched_getcpu(). rdpid is a 1 cycle instruction on Icelake servers. Benchmarks with ceph_test_c2c showed this was fractionally slower than THREAD_OWNER when # threads < # cores and fractionally faster than THREAD_OWNER when # threads > # cores.

I suspect that outside of benchmarks (where instructions/data are L1 cached) that SCHED_GETCPU USING RDPID will give the best performance, followed by THREAD_OTHER, followed by SCHED_GETCPU.